current_spring2026/retrieval/query_understanding.py

```
@@ -0,0 +1,180 @@
+
+    raw_json = response.choices[0].message.content.strip()
+    if raw_json.startswith("```"):
+        raw_json = raw_json.split("```")[1]
+        if raw_json.startswith("json"):
+            raw_json = raw_json[4:]
+        raw_json = raw_json.strip()
+
+    try:
+        parsed = json.loads(raw_json)
+    except json.JSONDecodeError as e:
+        raise ValueError(f"GPT-4o returned invalid JSON: {e}\nRaw: {raw_json}")
```

Ruff (0.15.9)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)